internal/dag/gatewayapi_processor_test.go

```diff
@@ -27,166 +27,251 @@ import (
+				fmt.Errorf("invalid hostname \"*.*.projectcontour.io\": [a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\\*\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"listener host & hostnames host do not exactly match": {
+			listenerHost: listenerHostname("listener.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
 			},
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("gateway hostname \"listener.projectcontour.io\" does not match hostnames hostname \"http.projectcontour.io\"")},
+		},
+		"listener host & hostnames host exactly match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & multi hostnames host exactly match one host": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+				"http2.projectcontour.io",
+				"http3.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: []error{
+				fmt.Errorf("gateway hostname \"http.projectcontour.io\" does not match hostnames hostname \"http2.projectcontour.io\""),
+				fmt.Errorf("gateway hostname \"http.projectcontour.io\" does not match hostnames hostname \"http3.projectcontour.io\""),
+			},
+		},
+		"listener host & hostnames host match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & hostnames host do not match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.example.com",
+			},
+			want:      map[string]struct{}{},
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"http.example.com\"")},
+		},
+		"listener host & wildcard hostnames host match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want: map[string]struct{}{
+				"http.projectcontour.io": {},
+			},
+			wantError: nil,
+		},
+		"listener host & wildcard hostname and matching hostname match": {
```

Comment:
similar to the above comment on this one

Reply:
My understanding is the listener.hostname defines the rules, so if it specifies a wildcard, and the route's hostname matches, then that route hostname is allowed to serve.